concordium-consensus/src/Concordium/TransactionVerification.hs

```diff
@@ -370,6 +370,94 @@ verifyNormalTransaction meta =
                         return $ Ok $ NormalTransactionSuccess (getHash keys) nonce
             )
 
+-- | Verifies an 'ExtendedTransaction' transaction.
+--  This function verifies the following:
+--  * Checks that enough energy is supplied for the transaction.
+--  * Checks that if a sponsor is specified, a sponsor signature is present as well.
+--  * CHecks that if no sponsor is specified, no sponsor signature is present.
+--  * Checks that the sender is a valid account.
+--  * Checks that the sponsor is a valid account, if present.
+--  * Checks that the nonce is correct.
+--  * Checks that the 'ExtendedTransaction' is correctly signed by both sender and the optional sponsor.
+verifyExtendedTransaction ::
+    forall m msg.
+    (TransactionVerifier m, Tx.TransactionData msg) =>
+    msg ->
+    m VerificationResult
+verifyExtendedTransaction meta =
+    either id id
+        <$> runExceptT
+            ( do
+                unless (Types.validatePayloadSize (Types.protocolVersion @(Types.MPV m)) (Tx.thPayloadSize (Tx.transactionHeader meta))) $
+                    throwError $
+                        NotOk InvalidPayloadSize
+
+                -- Check that either both, the sponsor and the sponsor signature are specified or neither.
+                mbSponsorAddrAndSig <- case (Tx.transactionSponsor meta, Tx.transactionSponsorSignature meta) of
+                    (Just sponsorAddr, Just sponsorSig) -> return $ Just (sponsorAddr, sponsorSig)
+                    (Nothing, Nothing) -> return Nothing
+                    (Just _sponsorAddr, Nothing) -> throwError $ NotOk SponsoredTransactionMissingSponsorSignature
+                    (Nothing, Just _sponsorSignature) -> throwError $ NotOk SponsoredTransactionMissingSponsor
+
+                -- Check that enough energy is supplied
+                let cost =
+                        Cost.baseCost
+                            (Tx.getTransactionHeaderPayloadSize $ Tx.transactionHeader meta)
+                            ( Tx.getTransactionNumSigs (Tx.transactionSignature meta)
+                                + maybe 0 Tx.getTransactionNumSigs (snd <$> mbSponsorAddrAndSig)
+                            )
```

Comment:
```suggestion
                let cost = Tx.transactionBaseCost meta
```
Note that the header size is actually different for an `ExtendedTransaction`, so using `getTransactionHeaderPayloadSize` will give the wrong result.